File: c/PE011.c
#include <stdio.h>

int main()
{
    freopen("PE011.txt", "r", stdin);
    int a[20][20];
    int i, j;
    for(i = 0; i < 20; i++)
        for(j = 0; j < 20; j++)
            scanf("%d", &a[i][j]);
    int mul, max = 0;
    for(i = 0; i < 20; i++)
        for(j = 0; j < 20; j++){
            if(i-3 >= 0 && j+3 < 20){
                mul = a[i][j]*a[i-1][j+1]*a[i-2][j+2]*a[i-3][j+3];
                if(mul > max)
                    max = mul;
            }
            if(j+3 < 20){
                mul = a[i][j]*a[i][j+1]*a[i][j+2]*a[i][j+3];
                if(mul > max)
                    max = mul;
            }
            if(i+3 < 20){
                mul = a[i][j]*a[i+1][j]*a[i+2][j]*a[i+3][j];
                if(mul > max)
                    max = mul;
            }
            if(i+3 < 20 && j+3 < 20){
                mul = a[i][j]*a[i+1][j+1]*a[i+2][j+2]*a[i+3][j+3];
                if(mul > max)
                    max = mul;
            }
        }
    printf("%d\n", max);
    return 0;
}

File: java/PE011.java
import java.util.Scanner;
import java.io.*;

public class PE011 {
    public static void main(String[] args) throws FileNotFoundException
    {
        int[][] m = new int[20][20];
        Scanner sc = new Scanner(new File("PE011.txt"));
        for(int i = 0; i < 20; i++) {
            for(int j = 0; j < 20; j++) {
                m[i][j] = sc.nextInt();
            }
        }
        int max = 0;
        int temp;
        for(int i = 0; i < 20; i++) {
            for(int j = 0; j < 20; j++) {
                if(isIn(i+3, j-3)) {
                    temp = m[i][j]*m[i+1][j-1]*m[i+2][j-2]*m[i+3][j-3];
                    if(temp > max)
                        max = temp;
                }
                if(isIn(i+3, j)) {
                    temp = m[i][j]*m[i+1][j]*m[i+2][j]*m[i+3][j];
                    if(temp > max)
                        max = temp;
                }
                if(isIn(i+3, j+3)) {
                    temp = m[i][j]*m[i+1][j+1]*m[i+2][j+2]*m[i+3][j+3];
                    if(temp > max)
                        max = temp;
                }
                if(isIn(i, j+3)) {
                    temp = m[i][j]*m[i][j+1]*m[i][j+2]*m[i][j+3];
                    if(temp > max)
                        max = temp;
                }
            }
        }
        System.out.println(max);
    }
    private static boolean isIn(int a, int b)
    {
        if((a >= 0 && a < 20) && (b >= 0 && b < 20))
            return true;
        return false;
    }
}
